opal/mca/common/ofi/common_ofi.c

```diff
@@ -446,63 +446,116 @@ static int check_provider_attr(struct fi_info *provider_info, struct fi_info *pr
-    if (NULL == topology) {
-        return false;
-    }
-
-    /* Allocate memory for proc_cpuset */
-    proc_cpuset = hwloc_bitmap_alloc();
-    if (NULL == proc_cpuset) {
-        return false;
-    }
-
-    /* Fill cpuset with the collection of cpu cores that the process runs on */
-    ret = hwloc_get_cpubind(topology, proc_cpuset, HWLOC_CPUBIND_PROCESS);
-    if (0 > ret) {
-        goto error;
-    }
-
-    /* Get the pci device from bdf */
-    obj = hwloc_get_pcidev_by_busid(topology, pci.domain_id, pci.bus_id, pci.device_id,
-                                    pci.function_id);
-    if (NULL == obj) {
-        goto error;
-    }
+	pmix_info_t *info, directive;
+	pmix_device_distance_t *distances, *nearest = NULL;
+	pmix_device_type_t type = PMIX_DEVTYPE_OPENFABRICS |
+	  PMIX_DEVTYPE_NETWORK |
+	  PMIX_DEVTYPE_COPROC |
+	  PMIX_DEVTYPE_GPU |
+	  PMIX_DEVTYPE_BLOCK |
+	  PMIX_DEVTYPE_DMA;
+
+	PMIX_INFO_LOAD(&directive, PMIX_OPTIONAL, NULL, PMIX_BOOL);
+	ret = PMIx_Get(&opal_process_info.myprocid,
+			 PMIX_DEVICE_DISTANCES, &directive, 1, &val);
+	PMIX_INFO_DESTRUCT(&directive);
+	if (ret != PMIX_SUCCESS) {
+		goto out;
+	}
+
+    if (PMIX_DATA_ARRAY != val->type) {
+        PMIX_VALUE_RELEASE(val);
+		goto release;
+    }
+    dptr = val->data.darray;
+    if (NULL == dptr) {
+        PMIX_VALUE_RELEASE(val);
+		goto release;
+    }
+    if (PMIX_DEVICE_DIST != dptr->type) {
+        PMIX_VALUE_RELEASE(val);
+		goto release;
+    }
+    distances = (pmix_device_distance_t*)dptr->array;
+
+	for (i = 0; i < dptr->size; i++)
+		fprintf(stderr, "%d: %d:%s:%d:%d\n", getpid(), i, distances[i].uuid,
+				distances[i].mindist, distances[i].maxdist);
+
+	nearest = calloc(sizeof(*distances), dptr->size);
+	if (!nearest)
+		goto release;
+
+	for (i = 0; i < dptr->size; i++) {
+		if (distances[i].mindist < near) {
+			idx = 0;
+			near = distances[i].mindist;
+			nearest[idx] = distances[i];
+			idx++;
+		} else if (distances[i].mindist == near) {
+			nearest[idx] = distances[i];
+			idx++;
+		}
+	}
+
+	*num_distances = idx;
+
+	goto out;
+
+release:
+	PMIX_VALUE_RELEASE(val);
+out:
+	return nearest;
+}
 
-    /* pcidev objects don't have cpusets so find the first non-io object above */
-    obj = hwloc_get_non_io_ancestor_obj(topology, obj);
-    if (NULL != obj) {
-        result = hwloc_bitmap_intersects(proc_cpuset, obj->cpuset);
-    }
+#if OPAL_OFI_PCI_DATA_AVAILABLE
 
-error:
-    hwloc_bitmap_free(proc_cpuset);
-    return result;
+static bool is_near(pmix_device_distance_t *distances,
+					int num_distances,
+					hwloc_topology_t topology,
+					struct fi_pci_attr pci)
+{
+	hwloc_obj_t pcidev, osdev;
+
+	pcidev = hwloc_get_pcidev_by_busid(topology, pci.domain_id,
+						pci.bus_id, pci.device_id,
+						pci.function_id);
+	if (!pcidev)
+		return false;
+
+	for(osdev = pcidev->io_first_child; osdev != NULL; osdev = osdev->next_sibling) {
+		int i;
+		const char *address = hwloc_obj_get_info_by_name(osdev, "Address");
```

Comment:
I think you lost something here - you need to address the two cases: (a) where the device is declared to be "network" (which has the "address" field), and (b) where the device is declared to be "openfabric" (which has "nodeguid" and "systemguid", IIRC).